Generic Guns rework

tenmillimaster opened the pull request:
#### Summary
```SUMMARY: Mods "Completely overhauls generic guns"```

#### Purpose of change
Generic Guns has been non-functional (to varying degrees) for a long, long time. It is difficult for content makers to maintain it, having to chase down multiple places for their work. This PR aims to close #35359 using the strategy outlined in #35427.

#### Describe the solution
Rather than having to copy and alter numerous items for the mod, Generic Guns will now utilize 'archetypal guns'(defined in #35483) that are defined in the mod, and are intended to stand on their own; thus, instead of carefully selecting which guns to black list, and which to not, **ALL MAINLINE** guns will get migrated instead to one of these archetypes.
In case  a gun doesn't adequately fit to one, a new archetype will be made.

The same approach is made for gunmods, magazines, ammunition, and recipes.


#### Describe alternatives you've considered
Just updating the blacklist.

#### Testing
Not much yet. I've loaded the game just prior, to make sure there aren't any load errors related to *this mod*. There is a feature this work is contingent on, which produces load errors, #35545
#### Additional context
This has been a HUGE project, and I have likely made mistakes. Any criticism, input, or suggestions are greatly appreciated.

As I was adding guns, I expanded the # of guns slightly to what I felt was a more complete list of archetypes.

tenmillimaster commented:
Current list of known issues:
Pipe guns needed to get added to migrations.
Some mainline ammo is still craftable.
None of the Generic guns ammo is craftable?

I-am-Erk commented:
So, I think we should make a CI check that makes sure all `type: ammo` and `type: guns` items are either migrated in Generic Guns, or are present on a whitelist to say they don't need migration.

I-am-Erk commented on data/mods/Generic_Guns/ammo/pistol_magnum.json:
```diff
@@ -0,0 +1,35 @@
+    "name": "magnum pistol ammo, ball",
+    "ammo_type": "ammo_pistol_magnum",
+    "casing": "pistol_magnum_casing",
+    "description": "Magnum pistol ammuniton, developed from early handgun hunting efforts and popularized by law enforcement usage.  It is more energetic than common pistol ammo.  'Ball' is an old term for full metal jacketed bullets which offer reduced fouling, better reliability, and slightly improved penetration compared to plain lead."
```
You said "energetic" here and "powerful" below. It's not a bit deal but you could go for consistency.
```suggestion
    "description": "Magnum pistol ammuniton, developed from early handgun hunting efforts and popularized by law enforcement usage.  It is more powerful than common pistol ammo.  'Ball' is an old term for full metal jacketed bullets which offer reduced fouling, better reliability, and slightly improved penetration compared to plain lead."
```

I-am-Erk commented on data/mods/Generic_Guns/firearms/black_powder.json:
```diff
@@ -0,0 +1,29 @@
+    "type": "GUN",
+    "name": "antique revolver",
+    "ammo": "ammo_black_powder",
+    "description": "This is an old revolver designed during the period of westward expansion. Black powder paper cartridges must be loaded for each shot, which is a fairly lengthy process.   Despite its age, this would perform adequately against most two-legged threats. ",
```
```suggestion
    "description": "This is an old revolver designed during the period of westward expansion.  Black powder paper cartridges must be loaded for each shot, which is a fairly lengthy process.   Despite its age, this type of weapon would perform adequately against most two-legged threats.",
```

I-am-Erk commented on data/mods/Generic_Guns/firearms/pistol.json:
```diff
@@ -0,0 +1,66 @@
+    "type": "GUN",
+    "name": "machine pistol",
+    "ammo": "ammo_pistol",
+    "description": "This pistol is a tiny machinegun you can stuff into a holster, with which you could dump its magazine at a blistering rate into any close range foes.  Machine pistols mostly see use by vehicle crewmen or bodygaurds of VIPs. Due to its preposterous rate of fire it is difficult to fire controllably.",
```
```suggestion
    "description": "This pistol is a tiny machinegun you can stuff into a holster, with which you could dump its magazine at a blistering rate into any close range foes.  Machine pistols mostly see use by vehicle crewmen or bodygaurds of VIPs.  Due to its preposterous rate of fire it is difficult to control.",
```

I-am-Erk commented on data/mods/Generic_Guns/firearms/pistol.json:
```diff
@@ -0,0 +1,66 @@
+    "type": "GUN",
+    "name": "police revolver",
+    "ammo": "ammo_pistol",
+    "description": "This may not be an actual police firearm, but similar revolvers chambered for standard defensive calibers like this were a favorite of police departments for nearly a century, up until the 1986 Miami shootout.  Afterwards, the slow reloading and shooting of revolvers were considered liabilities; still, this model's accuracy and moderate recoil make for a serviceable sidearm, a ...
```
```suggestion
    "description": "Revolvers like this, chambered for standard defensive calibers, were a favorite of police departments for nearly a century, up until the 1986 Miami shootout.  Afterwards, the slow reloading and shooting of revolvers were considered liabilities; still, this model's accuracy and moderate recoil make for a serviceable sidearm, and there are no magazines for you to lose or damage.",
```
YMMV but I think that's more concise

I-am-Erk commented on data/mods/Generic_Guns/firearms/rifle.json:
```diff
@@ -0,0 +1,56 @@
+    "type": "GUN",
+    "name": "service rifle",
+    "ammo": "ammo_rifle",
+    "description": "Borne from years of refinement .  The products of decades of improvement, rifle such as this are handy, reliable, and adaptable.   Short of large creatures and light vehicles, this should take care of most of your problems out to several hundred meters.",
```
```suggestion
	    "description": "Borne from years of refinement.  The products of decades of improvement, rifle such as this are handy, reliable, and adaptable.   Short of large creatures and light vehicles, this should take care of most of your problems out to several hundred meters.",
```

tenmillimaster replied:
@I-am-Erk oh right, I never finished this description...

Brian-Otten replied:
Just went ingame to check all the itemgroups. going to note down items that need migrating

luty smg .40
luty smg .45
luty smg 9mm
winchester m37
Browning auto 5
CZ-75
Glock 31
hi-point c9
Kel tec p32
Kel-tec p3AT
kel-tec pf-9
marlin 1895
mossberg 930 spx
PT 603
remington 1100 competition
remington 700
remington 870 express
ruger LCP
Sig p320
Walther p22
Walter p38
AT 4
m74 rocket clip
m110A1
Glock 18c
M17
m1911A1
Mk 23 MOD
mcmillan tac-50
mossberg 590A
remington 870 MCS
Magnum research BFR
Saiga-410
Taurus raging judge magnum
USAS 12
Uberti cattleman
Browning hi-power .40
Walther CCP
Walther PPQ
S&W model 10
Walther ppq 9mm
m45A1
Bond arms derringer
Ar-15
Hk417
1874 sharps
1887 bootleg shotgun
Colt lighting .45 carbine
m1987 trench gun
mac-11
Tavor TS12
IWI tavor x95
ar-10
Marlin 1895 SBL
Remington 700
Remington ACR .300
Cobray street sweeper
Franchi spas-12
PTR 603
H&K mp5k-pdw
H&K mpsd
.45-70 LFN
usas-12 drum magazine
saiga-410 drum magazine
usas-12 box magazine
saiga-410 box magazine
AF2011A1 magazine
.300ac blackout conversion kit
.410 000


ammo_shotgun_obscure spawns nothing
guns_launcher_grenade spawns nothing
guns_launcher_improvised spawns nothing
guns_smg_obscure spawns nothing
mags_shotgun_common spawns nothing

That's all the ones i could find easily, there might still be more.

tenmillimaster commented on data/mods/Generic_Guns/ammo/casings.json:
```diff
@@ -0,0 +1,72 @@
+    "copy-from": "50_casing",
+    "type": "GENERIC",
+    "name": "huge rifle casing",
+    "description": "A huge casing from an huge rifle round."
```
Poor attempt at humor.

Brian-Otten commented:
Just went ingame to check all the itemgroups. going to note down items that need migrating

luty smg .40
luty smg .45
luty smg 9mm
winchester m37
Browning auto 5
CZ-75
Glock 31
hi-point c9
Kel tec p32
Kel-tec p3AT
kel-tec pf-9
marlin 1895
mossberg 930 spx
PT 603
remington 1100 competition
remington 700
remington 870 express
ruger LCP
Sig p320
Walther p22
Walter p38
AT 4
m74 rocket clip
m110A1
Glock 18c
M17
m1911A1
Mk 23 MOD
mcmillan tac-50
mossberg 590A
remington 870 MCS
Magnum research BFR
Saiga-410
Taurus raging judge magnum
USAS 12
Uberti cattleman
Browning hi-power .40
Walther CCP
Walther PPQ
S&W model 10
Walther ppq 9mm
m45A1
Bond arms derringer
Ar-15
Hk417
1874 sharps
1887 bootleg shotgun
Colt lighting .45 carbine
m1987 trench gun
mac-11
Tavor TS12
IWI tavor x95
ar-10
Marlin 1895 SBL
Remington 700
Remington ACR .300
Cobray street sweeper
Franchi spas-12
PTR 603
H&K mp5k-pdw
H&K mpsd
.45-70 LFN
usas-12 drum magazine
saiga-410 drum magazine
usas-12 box magazine
saiga-410 box magazine
AF2011A1 magazine
.300ac blackout conversion kit
.410 000

ammo_shotgun_obscure spawns nothing
guns_launcher_grenade spawns nothing
guns_launcher_improvised spawns nothing
guns_smg_obscure spawns nothing
mags_shotgun_common spawns nothing

That's all the ones i could find easily, there might still be more.

esotericist commented:
> This apparently didn't work because the test itemgroup function in the debug menu does not properly take migrations into account. So this list is sadly mostly invalid.

migrations currently only happen on save game load, because that mechanism was originally only intended to be used for obsoleted items (and thus things that would no longer be in item lists). 

engineering is needed for this to work right.

See: #35545

tenmillimaster commented:
I have been playtesting somewhat with a compiled copy of #35545. Most everything seems to be in order, I missed a few migrations.

Currently I'm having issues with magazines- https://media.discordapp.net/attachments/598529174302490644/666523696155852800/unknown.png?width=349&height=197

they have a copy-from from a similar mainline mag, and I was under the impression redefining a field would override it for the new item. This wasn't happening- I tried "delete" : { ammo_type: "old magammotype"} as well, but that hasn't worked.

tenmillimaster commented:
I've gone and cleaned up some more of the recipes, threw in a very hacky hack to make vehicle turrets and their ammo spawns happy, for 0.E (until migration works properly with vehicle ammo).  There was a circular dependency error caused by 'bp_shot_scrap' and a recipe copying from it, related to the order in which they were defined, so I've gone and fixed that too.

I've added black powder versions of all the ammo, and de-referenced most of the recipes since copying and obsoleting wasn't going to work.

I've split all the pipe-weapons (pipe on a stick with a hammer) apart from the repeaters (somewhat well made firearms) in recipes, and have attempted to rationalize their difficulties and components.

I-am-Erk commented:
Well let's see how that goes
